oximeter/oximeter/src/quantile.rs

```diff
@@ -0,0 +1,393 @@
+    ) -> Self {
+        Self {
+            marker_hghts,
+            marker_pos,
+            desired_marker_pos,
+            desired_marker_incrs,
+        }
+    }
+
+    /// Construct a `Quantile` instance for the 50th/median percentile.
+    pub fn p50() -> Self {
+        Self::new(50).unwrap()
+    }
+
+    /// Construct a `Quantile` instance for the 90th percentile.
+    pub fn p90() -> Self {
+        Self::new(90).unwrap()
+    }
+
+    /// Construct a `Quantile` instance for the 95th percentile.
+    pub fn p95() -> Self {
+        Self::new(95).unwrap()
+    }
+
+    /// Construct a `Quantile` instance for the 99th percentile.
+    pub fn p99() -> Self {
+        Self::new(99).unwrap()
+    }
+
+    /// Get the p value as a float.
+    fn _p(&self) -> f64 {
+        self.desired_marker_incrs[2]
+    }
+
+    /// Get the p value as an integer.
+    pub fn p(&self) -> u64 {
+        (self.desired_marker_incrs[2] * 100.0) as u64
+    }
+
+    /// Estimate the p-quantile of the population.
+    ///
+    /// Returns an error if the sample is empty.
+    pub fn estimate(&self) -> Result<f64, QuantileError> {
+        // Return NaN if the sample is empty.
+        if self.is_empty() {
+            return Err(QuantileError::EmptyQuantile);
+        }
+
+        // Return the middle marker height if the sample size is at least 5.
+        if self.len() >= 5 {
+            return Ok(self.marker_hghts[2]);
+        }
+
+        let mut heights: [f64; 4] = [
+            self.marker_hghts[0],
+            self.marker_hghts[1],
+            self.marker_hghts[2],
+            self.marker_hghts[3],
+        ];
+
+        let len = self.len() as usize;
+        float_ord::sort(&mut heights[..len]);
+
+        let desired_index = (len as f64) * self._p() - 1.;
+        let mut index = desired_index.ceil();
+        if desired_index == index && index >= 0. {
+            let index = index.round_ties_even() as usize;
+            if index < len - 1 {
+                // `marker_hghts[index]` and `marker_hghts[index + 1]` are
+                // equally valid estimates,  by convention we take their average.
+                return Ok(0.5 * self.marker_hghts[index]
+                    + 0.5 * self.marker_hghts[index + 1]);
+            }
+        }
+        index = index.max(0.);
+        let mut index = index.round_ties_even() as usize;
+        index = min(index, len - 1);
+        Ok(self.marker_hghts[index])
+    }
+
+    /// Return the sample size.
+    pub fn len(&self) -> u64 {
+        self.marker_pos[4] as u64
+    }
+
+    /// Determine whether the sample is empty.
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Append a value/observation to the population and adjust the heights.
+    pub fn append<T>(&mut self, value: T) -> Result<(), QuantileError>
+    where
+        T: HistogramSupport,
+    {
+        if !value.is_finite() {
+            return Err(QuantileError::NonFiniteValue(format!("{:?}", value)));
```

Comment:
Nit, but I don't think we need to format the value. It's either -inf, +inf, or NaN, and I'm not sure knowing which would provide much.